docs/source/format/CProtocol.rst

```diff
@@ -0,0 +1,546 @@
+.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+.. KIND, either express or implied.  See the License for the
+.. specific language governing permissions and limitations
+.. under the License.
+
+=========================
+The Arrow C data protocol
+=========================
+
+Rationale
+=========
+
+Apache Arrow aims to be a universal in-memory format for the representation
+of tabular ("columnar") data, but some projects may face an undesirable
+choice between either depending on a fast-evolving dependency such as the
+Arrow C++ library, or having to implement adapters for data interchange
+(for example by reimplementing the Arrow IPC format, which is non-trivial
+and does not allow cross-runtime zero-copy).
+
+The Arrow C data protocol defines a very small, stable set of C definitions
+that can be easily *copied* in any project's source code and used for columnar
+data interchange in the Arrow format.  For non-C/C++ languages and runtimes,
+it should be almost as easy to translate the C definitions into the
+corresponding C FFI declarations.
+
+Applications and libraries can therefore choose between tight integration
+with the Arrow *software project* (benefitting from the growing array of
+facilities exposed by e.g. the C++ or Java implementations of Apache Arrow,
+but with the cost of a dependency) or minimal integration with the Arrow
+*format*.
+
+Goals
+-----
+
+* Expose an ABI-stable interface.
+* Easy for third-party projects to implement support for (including partial
+  support where sufficient), with little initial investment.
+* Zero-copy sharing of Arrow data between independent runtimes
+  and components running in the same process.
+* Match the Arrow array concepts closely, to avoid the development of
+  yet another marshalling layer.
+* Avoid the need for one-to-one adaptation layers such as the limited
+  JPype-based bridge between Java and Python.
+* Allow integration without an explicit dependency (either at compile-time
+  or runtime) on the Arrow software project.
+
+Ideally, the Arrow C data protocol can become a low-level *lingua franca*
+for sharing columnar data at runtime and establish Arrow as the universal
+building block in the columnar processing ecosystem.
+
+Non-goals
+---------
+
+* Expose a C API mimicking operations available in higher-level runtimes
+  (such as C++, Java...).
+* Data sharing between distinct processes.
+
+Comparison with the Arrow IPC format
+------------------------------------
+
+Pros of the C data protocol vs. the IPC format:
+
+* No dependency on Flatbuffers.
+* No buffer reassembly (data is already exposed in logical Arrow format).
+* Zero-copy by design.
+* Easy to reimplement from scratch.
+
+Pros of the IPC format vs. the data protocol:
+
+* Works accross processes and machines.
+* Allows data storage and persistency.
+* Being a streamable format, has room for composing more features (such as
+  integrity checks, compression...).
```

Comment:
compression/encoding should probably be accounted for in this spec.  Perhaps its another struct or another member.  Since it doesn't exist yet it seems OK to wait on this.

Reply:
Do you think that could fit in the `flag` field? Perhaps I should make the flags a `int64_t`?

Reply:
Yeah, sorry just getting to that int64_t would be better there I think.

Reply:
thinking about it some more flags are a partial solution I think.  We would still need a way to describe buffer encodings separately (a strawman would be a separate buffer description string using the same format below, one for each buffer).  But this is conjecture, I think we can update the spec once we get agreement on  the encoding proposal (hopefully before the next release) and update the specification at that point.

Reply:
Ideally the struct remains ABI-stable, so frozen in its definition. So it would have to be added as e.g. a metadata entry, which is perhaps not the most immediate way to access it.

Reply:
Agreed, I was thinking we would have until 1.0 release to have it finalized.  If we need to make a decision now I think we can go one of two ways:
1. adding a flag below value for "alternative buffer encodings" and then we can add a formal definition to metadata to describe any additional encodings that are approved.  
2.  Add an additional optional `char*` field to describe encodings that is left undefined for the time being.